Top down `EnforceSorting`, Extended testbench for `EnforceSorting` rule to prepare for refactors, additional functionality such as pushdowns over unions

mustafasrepo opened the pull request:
# Which issue does this PR close?

N/A

# Rationale for this change

As some of you may know, we have been (and still are) collaborating with @mingmwang  on a top-down refactoring of the `EnforceSorting` [rule](https://github.com/apache/arrow-datafusion/pull/5290).  During our quest to find the simplest implementation that still offers full functionality, we observed that bottom-up and top-down approaches have their pros and cons. A bottom-up approach is better avoiding pipeline breakages, while a top-down approach results in simpler code and is able to perform certain push down operations its bottom-up sibling can not.

It is still not entirely clear whether we can achieve full functionality in a pure top-down approach, but in the interim, we can combine the two approaches to offer full functionality AND implement a test suite to guide our future efforts.

# What changes are included in this PR?
This PR implements a hybrid top-down/bottom-up sort optimization rule by utilizing the existing bottom-up approach and the top-down approach implemented by @mingmwang . With this hybrid rule, all test plans either improve or stay the same. We are also checking in a much larger test suite verifying all sorts of improvements like union pushdowns etc. We think the new test suite will serve as a very useful baseline for future improvements on the `EnforceSorting` rule (including our ongoing collaboration on finding whether a pure top-down approach is possible).

# Are these changes tested?

New unit tests are added. Existing tests also pass.

# Are there any user-facing changes?

No.

mustafasrepo commented on datafusion/core/src/physical_optimizer/sort_enforcement.rs:
```diff
@@ -146,19 +151,11 @@ impl PlanWithCorrespondingSort {
                     return None;
                 }
                 let is_spm = is_sort_preserving_merge(plan);
-                let is_union = plan.as_any().is::<UnionExec>();
```
Since we now have a pushdown subroutine. We no longer need special handling for union here.

alamb replied:
I like the idea of separting the "enforce sort requirements" from the "optimize sort requirements" which I think this PR improves

mustafasrepo commented on datafusion/core/src/physical_optimizer/sort_enforcement.rs:
```diff
@@ -246,11 +243,11 @@ impl PlanWithCorrespondingCoalescePartitions {
                 // maintain a single partition. If we just saw a `CoalescePartitionsExec`
                 // operator, we reset the tree and start accumulating.
                 let plan = item.plan;
-                if plan.as_any().is::<CoalescePartitionsExec>() {
```
I have just swapped `if` and `else if` to make it more readable.

mustafasrepo commented on datafusion/core/src/physical_optimizer/sort_enforcement.rs:
```diff
@@ -343,77 +347,60 @@ impl PhysicalOptimizerRule for EnforceSorting {
-                coalesce_onwards: vec![None],
-            }));
-        }
+    if plan.children().is_empty() || coalesce_onwards[0].is_none() {
```
I have just removed nested `if` by early returning in `is_none` case.

mustafasrepo commented on datafusion/core/src/physical_optimizer/sort_enforcement.rs:
```diff
@@ -439,62 +426,64 @@ fn ensure_sorting(
-                if !plan.maintains_input_order()[idx] {
-                    let count = plan.output_ordering().map_or(0, |e| e.len());
-                    if (count > 0) && !is_sort(&plan) {
-                        update_child_to_change_finer_sort(child, sort_onwards, count)?;
```
Because of pushdown routine we no longer need to change below sorts with finer sorts. This handling can be removed from here.

mustafasrepo commented on datafusion/core/src/physical_optimizer/sort_enforcement.rs:
```diff
@@ -1101,22 +1064,21 @@ mod tests {
 
-        // let filter_exec = sort_exec;
-        let physical_plan = window_exec("non_nullable_col", sort_exprs, filter);
+        let physical_plan = bounded_window_exec("non_nullable_col", sort_exprs, filter);
```
I have changed the `window_exec` with `bounded_window_exec` because the window expression used in the expression can run with `BoundedWindowAggExec` (If it were come from sql query it would work with `BoundedWindowExec`). I just want it to be consistent with general behaviour.

mustafasrepo commented on datafusion/core/src/physical_optimizer/sort_pushdown.rs:
```diff
@@ -0,0 +1,394 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
This file is pruned version of top-down rule by @mingmwang at #5290. Its scope is converted from top-down sort enforcement to sort pushdown.

mustafasrepo commented on datafusion/core/tests/sql/select.rs:
```diff
@@ -841,7 +841,7 @@ async fn sort_on_window_null_string() -> Result<()> {
     ])
     .unwrap();
 
-    let ctx = SessionContext::with_config(SessionConfig::new().with_target_partitions(2));
```
Result of these tests were not deterministic. I have changed so that result would become deterministic.

mustafasrepo commented on datafusion/core/tests/window_fuzz.rs:
```diff
@@ -329,9 +329,10 @@ async fn run_window_test(
-    let exec1 = Arc::new(
-        MemoryExec::try_new(&[vec![concat_input_record]], schema.clone(), None).unwrap(),
-    );
+    let memory_exec =
```
With new changes `MemoryExec` now supports `output_ordering`. Since in these tests input is sorted. I give this information to `MemoryExec`.

ozankabak commented:
@alamb, when you have some time, can you check this out? We have some more cool improvements coming in depending on this (sneak peek: streaming aggregations (i.e. streaming GROUP BY), non-pipeline breaking window queries involving PARTITION BY clauses)

alamb commented:
Starting to check this one out. Sorry I was thinking (hoping 😆  ) that @mingmwang  would have a chance to review too

alamb commented:
I am not going to get to this today but I will get to it tomorrow

ozankabak commented:
Thank you!

alamb commented on datafusion/core/src/physical_optimizer/dist_enforcement.rs:
```diff
@@ -2104,7 +2080,6 @@ mod tests {
         // The optimizer should not add an additional SortExec as the
         // data is already sorted
         let expected = &[
```
Why was the `SortPreservingMergeExec` removed? Is it because there is a single partition in the input? 

I think the intent of this test was to ensure when a `SortPreservingMergeExec` a new sort was not added.

Would it be possible to update the test so that the `SortPreservingMergeExec` is still present (perhaps by increasing the number of input partitions)? 

Perhaps using the same strategy of `parquet_exec_multiple_sorted`

mustafasrepo replied:
Thanks for pointing this out. I have converted input source to `parquet_exec_multiple_sorted` test now keeps `SortPreservingMergeExec`.

alamb commented on datafusion/physical-expr/src/sort_expr.rs:
```diff
@@ -69,4 +62,70 @@ impl PhysicalSortExpr {
+/// Represents sort requirement associated with a plan
+#[derive(Clone, Debug)]
+pub struct PhysicalSortRequirement {
+    /// Physical expression representing the column to sort
```
Is expr always a `PhysicalSortExpr`? If so the options on this structure might be redundant with the options on another

https://docs.rs/datafusion/latest/datafusion/physical_plan/expressions/struct.PhysicalSortExpr.html

ozankabak replied:
The main difference between a `PhysicalSortExpr` and a `PhysicalSortRequirement` is that `SortOptions` is optional in the latter, but not the former.

The former carries the actual ordering information in the plan, while the latter is used to specify ordering requirements. A natural question here would be: Why do we need two structs? Can't we just use `PhysicalSortExpr` in both contexts? That is indeed what we used to do, but our collaboration with @mingmwang revealed that there are cases where a requirement specifies the sorting columns, but doesn't care about the ascending/descending distinction. Therefore, he proposed the `PhysicalSortRequirement` in one of his commits. We agreed with the rationale and adopted it in this work.

I hope that helps!

alamb commented on datafusion/physical-expr/src/sort_expr.rs:
```diff
@@ -69,4 +62,70 @@ impl PhysicalSortExpr {
+
+/// Returns the SQL string representation of the given [SortOptions] object.
+#[inline]
+fn to_string(options: &SortOptions) -> &str {
```
This might be a nice thing (`impl Display`) to contribute upstream to arrow-rs eventually

alamb commented on datafusion/physical-expr/src/sort_expr.rs:
```diff
@@ -69,4 +62,70 @@ impl PhysicalSortExpr {
+
+/// Returns the SQL string representation of the given [SortOptions] object.
+#[inline]
+fn to_string(options: &SortOptions) -> &str {
```
```suggestion
fn to_str(options: &SortOptions) -> &str {
```

Given it returns a `str` I think calling it to `to_string` is slightly misleading

ozankabak replied:
Agreed 👍

alamb commented on datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs:
```diff
@@ -123,7 +122,7 @@ impl BoundedWindowAggExec {
         // All window exprs have the same partition by, so we just use the first one:
         let partition_by = self.window_expr()[0].partition_by();
-        let sort_keys = self.sort_keys.as_deref().unwrap_or(&[]);
+        let sort_keys = self.input.output_ordering().unwrap_or(&[]);
```
this makes sense to me -- specifically that `BoundedWindowExec` does not change the ordering of its input and thus there is no reason to store the sort key directly

alamb reviewed (commented):
Thank you for this contribution @mustafasrepo .

I have not yet reviewed the changes in `datafusion/core/src/physical_optimizer/sort_enforcement.rs` and `datafusion/core/src/physical_optimizer/sort_pushdown.rs` -- I will try and find time to do so later. 

I would like to again encourage you to break up your PRs into smaller chunks. I know it is more effort to create multiple smaller PRs, but I think it will have benefits:

1. The quality of the reviews will likely be higher: a) more time (in aggregate) can be devoted by crowdsourcing from multiple reviewers b)  we are more likely to be able to find someone with more specialized knowledge of a smaller area to provide better feedback c) It will be easier to have a targeted discussion about certain changes.
2. The latency between filing and review should be lower
3. The potential for accumulating conflicts is often lower

For example, in this PR there are several smaller improvements that seem like they could be broken out individually:
1. various cleanups in `datafusion/physical-expr/src/utils.rs`, 
2. The removal of sort key calculation from SymmetricHashJon, WindowExec and BoundedWindowExec, and the physical planer
4. The addition of `PhysicalSortRequirement`

I know in the past @ozankabak has made the point that more smaller PRs can often make it hard to understand the rationale, especially for larger or more connected changes. 


One way to help with that concern is to make a draft PR with all the large changes and refer to that in the smaller PR that

ozankabak commented:
> One way to help with that concern is to make a draft PR with all the large changes and refer to that in the smaller PR

This a good idea! We will try this new pattern in our follow-up work (i.e. non-pipeline-breaking PARTITION BYs in windows)

alamb commented on datafusion/core/src/physical_plan/joins/symmetric_hash_join.rs:
```diff
@@ -399,10 +393,14 @@ impl ExecutionPlan for SymmetricHashJoinExec {
     }
 
-    fn required_input_ordering(&self) -> Vec<Option<&[PhysicalSortExpr]>> {
+    fn required_input_ordering(&self) -> Vec<Option<Vec<PhysicalSortRequirement>>> {
```
I actually spent some time trying to test bringing this change to IOx and I found this particular part quite challenging:
1. Figuring out `Vec<Option<Vec<..>>>` and how to 
2. `make_sort_requirements_from_exprs` appears not to be public so I couldn't use it.

If we are going to change the signature of these traits, I think we should think about how make it easier to use.

For example,  I wonder if we could make a structure like `Distribution` to https://docs.rs/datafusion/20.0.0/datafusion/physical_plan/enum.Distribution.html to describe an `Order`? 


This would both 
1. make it more convenient to construct these lists 
2. provide a place to document what they are and all the subtlety 

Perhaps something like 

```rust
impl ExecutionPlan {
    // the order required for the plans inputs
    fn required_input_ordering(&self) -> Vec<Ordering> ;

    // output order produced by this plan
    fn output_ordering(&self) -> Ordering ;
}
```

And then encapsulate all the gory details about orderings and comparing them in `Ordering`:

```rust 
struct Ordering {
  exprs: Vec<dyn PhysicalExpr>
  // parallel list if there are any particular requirements on the sort orders
  required_options: Vec<SortOptions>
}

impl RequiredInputOrdering {
  // construct a required input 
  pub fn from_sort_exprs(exprs: Vec<dyn PhysicalSortExpr>) -> Self {
  .. 
  }
 /// does this represent any requirement on the input ordering
 fn input_order_required(&self) ->  bool {}

 // Is there any NULLS FIRST / NULLS LAST 
...
}
```

ozankabak replied:
I think you mean something like:
```rust
type Ordering = Vec<PhysicalSortExpr>;
type OrderingRequirement = Vec<PhysicalSortRequirement>;
```
and which would allow us increase readability by changing
```rust
impl ExecutionPlan {
    fn required_input_ordering(&self) -> Vec<Option<Vec<PhysicalSortRequirement>>>;
    fn output_ordering(&self) -> Option<&[PhysicalSortExpr]>;
}
```
to
```rust
impl ExecutionPlan {
    fn required_input_ordering(&self) -> Vec<Option<OrderingRequirement>>;
    fn output_ordering(&self) -> Option<Ordering>;
}
```
Am I understanding your idea right?

ozankabak replied:
We can also create new outright types instead of using aliases, as in:
```rust
struct Ordering(Vec<PhysicalSortExpr>);
struct OrderingRequirement(Vec<PhysicalSortRequirement>);
```
We would need some boilerplate to make them ergonomic when using elsewhere; i.e.
```rust
impl core::ops::Deref for Ordering {
    type Target = Vec<PhysicalSortExpr>;

    fn deref (self: &'_ Self) -> &'_ Self::Target {
        &self.0
    }
}

impl core::ops::DerefMut for Ordering {
    fn deref_mut (self: &'_ mut Self) -> &'_ mut Self::Target {
        &mut self.0
    }
}

impl core::ops::Deref for OrderingRequirement {
    type Target = Vec<PhysicalSortRequirement>;

    fn deref (self: &'_ Self) -> &'_ Self::Target {
        &self.0
    }
}

impl core::ops::DerefMut for OrderingRequirement {
    fn deref_mut (self: &'_ mut Self) -> &'_ mut Self::Target {
        &mut self.0
    }
}
```
which will allow us to avoid a public `make_sort_requirements_from_exprs` and instead offer:
```rust
impl From<Ordering> for OrderingRequirement {
    fn from(value: Ordering) -> Self {
        OrderingRequirement(make_sort_requirements_from_exprs(&value))
    }    
}
```

If I get the spirit of your suggestion right, we can experiment with these and incorporate such a change in our next commit addressing the feedback. We can explore aliasing vs. tuple-struct methods and share our findings along with the commit.

ozankabak replied:
OK, having played with these ideas for 2-3 hours, here are my observations:

Changing the return types of `output_ordering` and `required_input_ordering` with non-plug-in-compatible types are really challenging. I made two refactor attempts: One with a new `struct` and one with a tuple-struct like my previous example. I couldn't finish either of them in 2-3 hours -- it turns out doing this results in massive changes and things are quite intertwined.

However, I have some good news too: We can achieve an almost-as-good readability level with type aliasing:
```rust
pub type ExprOrdering = Vec<PhysicalSortExpr>;
pub type ExprOrderingRef<'a> = &'a [PhysicalSortExpr];
pub type OrderingRequirement = Vec<PhysicalSortRequirement>;
```
With these "types", I was able complete the refactor and significantly simplify how the code reads; i.e.
```rust
impl ExecutionPlan {
    fn required_input_ordering(&self) -> Vec<Option<OrderingRequirement>>;
    fn output_ordering(&self) -> Option<ExprOrderingRef>;
}
```
with `ExprOrdering` replacing `Vec<PhysicalSortExpr>` where the actual vector is required.

This refactor is not large in terms of LOCs (~250 lines change, almost entirely consisting of simple substitutions), but it touches almost 50 files! So we may want to leave it to a follow-on too -- I am fine with either.

ozankabak commented on datafusion/physical-expr/src/utils.rs:
```diff
@@ -115,33 +118,21 @@ pub fn normalize_out_expr_with_alias_schema(
-                    None => None,
-                };
-            Ok(normalized_form)
+            Ok(match expr.as_any().downcast_ref::<Column>() {
```
Right

ozankabak commented on datafusion/physical-expr/src/utils.rs:
```diff
@@ -210,31 +219,176 @@ pub fn ordering_satisfy_concrete<F: FnOnce() -> EquivalenceProperties>(
     }
 }
 
+pub fn map_columns_before_projection(
```
This was a pre-existing function that just got moved, but we will add comments to help future readers.

mingmwang commented:
I have time this week and will this PR carefully.

alamb commented:
I suggest we split this PR into two  
1. The changes to the type signature of `ExecutionPlan` which while large is mostly mechanical
2. The algorithm changes to `datafusion/core/src/physical_optimizer/sort_enforcement.rs` and `datafusion/core/src/physical_optimizer/sort_pushdown.rs`

I think we can get 1 in sooner -- I think after @mingmwang 's review of the algorithm we'll be ready for 2.

alamb commented:
I am sorry -- I am really struggling to find the time to review this PR thoroughly

ozankabak commented:
No worries, take your time. I will send the type aliases in a follow on PR. @mustafasrepo, can you make a new PR for (1)? Let's get that merged and this PR will be (2).

We will also try the strategy you suggested for the upcoming feature PRs.

mustafasrepo commented:
> I suggest we split this PR into two
> 
> 1. The changes to the type signature of `ExecutionPlan` which while large is mostly mechanical
> 2. The algorithm changes to `datafusion/core/src/physical_optimizer/sort_enforcement.rs` and `datafusion/core/src/physical_optimizer/sort_pushdown.rs`
> 
> I think we can get 1 in sooner -- I think after @mingmwang 's review of the algorithm we'll be ready for 2.

I have moved the changes related signature of `ExecutionPlan` to another PR. New PR can be found in #5772. Thanks @alamb for the suggestion.

mingmwang commented:
@mustafasrepo 
Could you please add the `global sort `flag back in the plan Display?  I think it is useful.

mustafasrepo commented:
> @mustafasrepo Could you please add the `global sort `flag back in the plan Display? I think it is useful.

Since existing tests also changes with this change, it increases diff a bit. If it is OK for you, I can file another Pr for this change. I guess it will be merged faster that way also. Like we did in #5772. What do you think?

ozankabak commented:
> @mustafasrepo Could you please add the `global sort `flag back in the plan Display? I think it is useful.

Yes, let's do it in a follow on PR just that only does that (with no other test changes) for ease of review and avoiding simple mistakes.

mingmwang commented:
> > @mustafasrepo Could you please add the `global sort `flag back in the plan Display? I think it is useful.
> 
> Yes, let's do it in a follow on PR after this, making only that change (with no other test changes) for ease of review and avoiding simple mistakes.

Agree.

ozankabak commented:
@alamb or @mingmwang: Given that the API PR has merged, let us know when you are comfortable to go ahead with this. After this merges, we will create two follow-ons: (1) fixing the type aliases, (2) printing global sorting information.

alamb commented:
Thanks for the ping @ozankabak  -- I will try and find time to review this PR carefully tomorrow

alamb commented on datafusion/core/src/physical_plan/joins/sort_merge_join.rs:
```diff
@@ -249,6 +249,17 @@ impl ExecutionPlan for SortMergeJoinExec {
+    fn maintains_input_order(&self) -> Vec<bool> {
+        match self.join_type {
+            JoinType::Inner => vec![true, true],
+            JoinType::Left | JoinType::LeftSemi | JoinType::LeftAnti => vec![true, false],
```
the `false` here is because NULLs can be injected into the stream arbitrarily, right, messing up the sort order?

mustafasrepo replied:
Exactly.

mingmwang replied:
👍

alamb commented on datafusion/core/src/physical_optimizer/sort_enforcement.rs:
```diff
@@ -1432,22 +1514,22 @@ mod tests {
             "SortPreservingMergeExec: [nullable_col@0 ASC,non_nullable_col@1 ASC]",
-            "  SortExec: expr=[nullable_col@0 ASC,non_nullable_col@1 ASC]",
-            "    UnionExec",
+            "  UnionExec",
```
This plan seems to have gone "in reverse" (as in the test says that the sort should be removed) -- is that ok?

ozankabak replied:
Yes, all looks good here. We move sorts down whenever doing so can result in sorting smaller datasets. In this case, having the sort above `UnionExec` or below doesn't matter in that regard, and the default tendency is to push sorts down in such cases.

alamb commented on datafusion/core/src/physical_optimizer/sort_enforcement.rs:
```diff
@@ -1669,25 +1842,378 @@ mod tests {
+        let physical_plan =
+            bounded_window_exec("non_nullable_col", sort_exprs1, window_agg2);
+
+        let expected_input = vec![
```
All of these tests must have taken substantial amounts of time to write and updated the expected output. 

One thing we started doing in IOx was to use  https://insta.rs/ (where basically you can do `cargo insta review` and the tool will update the inlined expected output for you). I don't have time to lead another refactoring project now (I am still trying to finish up sqllogictests) but using insta might help accelerate velocity

ozankabak replied:
Thanks for the pointer, we will take a look at it 👍

alamb reviewed (approved):
I went through the plan changes carefully -- they look reasonable to me. It feels like it would be good to make a physical "analyzer" pass (analogous to what is being done to the logical plans) to make the split between "update the plan to be correct" and "make the plan faster"

Thank you for sticking with this one @mustafasrepo  -- it looks like a real improvement to me (and the new tests are 👌 )

mingmwang commented:
@mustafasrepo @ozankabak 
My original plan was try to leverage the top-down process to achieve two goals:
1) Simply the optimization process, make it easily to understand
2) Can do all the sort optimizations in one pass, 
But when I start the implementation and testing I realize that it is quite hard to achieve the two goals at same time.  I think the new implementation is the hybrid of bottom-up and top-down approach but it is more complex. 
I accept that in the current phase, we combine the two approaches to offer full functionality. But in future I think we should still pursue the goal to offer the full functionality in a pure top-down approach.

ozankabak commented:
> But in future I think we should still pursue the goal to offer the full functionality in a pure top-down approach.

Agreed, let's keep on collaborating to figure out how we can accomplish that. Now that we have extensive tests, it should be easier to try our hand at doing that

mingmwang commented on datafusion/core/src/physical_optimizer/utils.rs:
```diff
@@ -68,3 +74,40 @@ pub fn add_sort_above(
     }
     Ok(())
 }
+
+/// Checks whether the given executor is a limit;
+/// i.e. either a [`LocalLimitExec`] or a [`GlobalLimitExec`].
+pub fn is_limit(plan: &Arc<dyn ExecutionPlan>) -> bool {
+    plan.as_any().is::<GlobalLimitExec>() || plan.as_any().is::<LocalLimitExec>()
+}
+
```
I would suggest avoid using the word  `executor`, maybe you can use input `plan` or input `operator` instead. 
Executor usually refers to different things, for example in Flink it is the TaskManager, in Spark or Ballista it is the Worker who runs the tasks(physical plan fragments).

mingmwang commented:
Overall, this PR LGTM. 
@alamb @ozankabak Could you please leave it for one more day before merge this ?

ozankabak commented:
Sure! Thanks for reviewing.

mingmwang commented on datafusion/physical-expr/src/utils.rs:
```diff
@@ -312,6 +305,91 @@ pub fn ordering_satisfy_requirement_concrete<F: FnOnce() -> EquivalencePropertie
+    }
+}
+
+/// This function maps back requirement after ProjectionExec
+/// to the Executor for its input.
+// Specifically, `ProjectionExec` changes index of `Column`s in the schema of its input executor.
+// This function changes requirement given according to ProjectionExec schema to the requirement
+// according to schema of input executor to the ProjectionExec.
+// For instance, Column{"a", 0} would turn to Column{"a", 1}. Please note that this function assumes that
+// name of the Column is unique. If we have a requirement such that Column{"a", 0}, Column{"a", 1}.
+// This function will produce incorrect result (It will only emit single Column as a result).
+pub fn map_columns_before_projection(
```
Nice comment !

mingmwang commented on datafusion/core/src/physical_optimizer/sort_pushdown.rs:
```diff
@@ -0,0 +1,416 @@
+        Ok(VisitRecursion::Continue)
+    }
+
+    fn map_children<F>(mut self, transform: F) -> Result<Self>
```
The original `self` in `map_children` is immutable, why it was changed to be mutable here ?

```rust
fn map_children<F>(self, transform: F) -> Result<Self>
```

ozankabak replied:
Using a `mut` binding enables us to reuse the same object with a simple attribute assignment without creating a new object.

mingmwang commented on datafusion/core/src/physical_optimizer/sort_pushdown.rs:
```diff
@@ -0,0 +1,416 @@
+        let required_ordering = new_plan
+            .output_ordering()
+            .map(make_sort_requirements_from_exprs);
+        // Since new_plan is a SortExec, we can safely get the 0th index.
+        let child = &new_plan.children()[0];
+        if let Some(adjusted) =
+            pushdown_requirement_to_children(child, required_ordering.as_deref())?
+        {
+            // Can push down requirements
+            Ok(Transformed::Yes(SortPushDown {
+                plan: child.clone(),
+                required_ordering,
+                adjusted_request_ordering: adjusted,
+            }))
+        } else {
+            // Can not push down requirements
+            Ok(Transformed::Yes(SortPushDown::init(new_plan)))
+        }
```
I have one question for this logic, since the parent requirements are already satisfied by modifying the SortExec, why you need to push down the requirements here ?

mustafasrepo replied:
Actually, you are right. No need for it. Changed it to `None`.

mingmwang commented on datafusion/core/src/physical_optimizer/sort_pushdown.rs:
```diff
@@ -0,0 +1,416 @@
+    {
+        // If the current plan is a leaf node or can not maintain any of the input ordering, can not pushed down requirements.
+        // For RepartitionExec, we always choose to not push down the sort requirements even the RepartitionExec(input_partition=1) could maintain input ordering.
+        // For RepartitionExec, we always choose to not push down the sort requirements even the RepartitionExec(input_partition=1) could maintain input ordering.
```
Duplicate comment here.

```rust
// For RepartitionExec, we always choose to not push down the sort requirements even the RepartitionExec(input_partition=1) could maintain input ordering

```

mingmwang commented on datafusion/core/src/physical_optimizer/sort_pushdown.rs:
```diff
@@ -0,0 +1,416 @@
+    plan: &Arc<dyn ExecutionPlan>,
+    parent_required: Option<&[PhysicalSortRequirement]>,
+) -> Result<Option<Vec<Option<Vec<PhysicalSortRequirement>>>>> {
+    const ERR_MSG: &str = "Expects parent requirement to contain something";
```
This structure is not easy to read and understand. I think we may need some future refactoring on this.

`Result<Option<Vec<Option<Vec<PhysicalSortRequirement>>>>>`

ozankabak replied:
This will be somewhat simplified when we use the type aliases we tested above. But I agree that we will probably simplify/refactor this code over time

mingmwang commented:
@ozankabak @alamb 
Please go ahead and merge this PR.

ozankabak commented:
@alamb, we addressed the reviews and this is good to go from our perspective

alamb commented:
Thanks again everyone!
